netket/vqs/base.py

```diff
@@ -316,13 +316,14 @@ def expect(vstate: VariationalState, operator: AbstractOperator):
 
 
 # default dispatch where use_covariance is not specified
-@dispatch
+@dispatch(precedence=10)
```

Comment:
Why is the precedence needed here? And could you maybe add a comment explaining it? With the dispatch, related code can be a bit hard find anyways (because it can be everywhere) and adding a precedence aggravates that further.

Reply:
@PhilipVinc ?

Reply:
I added an extensive comment explaining it.
This is not needed, but it makes the dispatch logic more user-friendly (for users who are not familiar to dispatch).

In general, if an user defines
```
def expect_and_grad(vs: MCState, op: MyNiceOperator, use_covariance: Any):
   ...
```
There is an ambiguity because the method defined above has stricter signatures for `vs` and `op` but weaker signature for `use_covariance` compared to the function
```python
def expect_and_grad(vs: VariationalState, op: AbstractOperator, use_covariance: bool):
```
This is called _signature ambiguity_ because the resolver does not know which to favour. 
In general the error would be fixed by the user declaring `use_covariance: bool`, but to be more user-friendly, we just let users be not-so-precise and fix this by setting higher precedence to our own rule.

Reply:
Okay, thanks for adding the comment. For my understanding: The resolver will now favor the the `VariationalState`, `AbstractOperator`, `bool` version over `MCState`, `CustomOp`, `Any`, right? But then, how is the user's version getting called at all in that case?

Reply:
Because the default method changes the type of `use_covariance` from `bool` to `Bool` (aka, `Val[True]()` or `Val[False]()`)
which are not `bool`, so it never calls itself.

This is the main reason why this method is there.

Reply:
Basically, when you call `expect_and_grad(vstate, op, None)` or `expect_and_grad(vstate, op, true)` this default method converts the boolean to `Val[True]` or `Val[False]` (which i call `TrueT` or `FalseT` in netket. 
This is a trick to allow dispatch based on a value (not type) because I want to dispatch to different methods whever i want to use the covariance formula or not.

Reply:
This could in principle be removed.
Probably would be a good idea to remove it, and just define two different functions: `expect_and_grad_covariance` and `expect_and_grad_nocovariance`

Reply:
I admit there is a certain elegance to this, but personally I'd prefer having `expect_and_grad_covariance` and `expect_and_grad_nocovariance` or so. That is clearer, especially in a Python context.

Also, I still don't see at what point the user code would then be called, as it would also not match either of the `TrueT`/`FalseT` variants.

Reply:
`Any` and `Bool` are supertypes of `TrueT` and `FalseT`, so user code would get called.

Anyhow, I agree that this part is confusing and I never liked this TrueT/FalseT trick, and I would like to get rid of it (for example with the two methods as proposed above), but that is for a later PR/time.